### PR TITLE
chore: avoid the risk of collisions in `CidHashSet`

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Forest is dealing with such a small number of CIDs (tens of millions) that they could be stored as 64 bit numbers with only 1 expected collision every 1000 years or so. But let's not risk it. Storing the full hash is only a few bytes more.

Changes introduced in this pull request:
- Store the full 32-byte hash for the most commonly used CID variant.

All benchmarks were run on storage-optimized droplets with 16GiB of RAM and 300 GiB of SSD. The average of at least three runs were used:

|   | Time (seconds)  | Peak RSS (GiB)  |
|---|---|---|
| Export snapshot (baseline) |  4978 (x1.00) | 4.81 (x1.00) |
| Export snapshot (BTreeSet)  | 5500 (x1.10)  | 6.69 (x1.39)  |
| Export snapshot (HashSet)  | 5331 (x1.07)  |  6.07 (x1.26) |

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by lemmih** [comment]
> How abt making fallback a variant of CidVariant?

I meant for the `CidVariant` enum to be the set of known CID types that we want to treat differently. This way, it always consists of a tag and an array of bytes, and it is smaller in size than a `Cid`.

**Comment by LesnyRumcajs** [comment]
This seems like a significant drop in performance. Are you positive we are fine with this tradeoff? What would be the result of a collision, in practice?

**Comment by lemmih** [comment]
> This seems like a significant drop in performance. Are you positive we are fine with this tradeoff? What would be the result of a collision, in practice?

A collision during snapshot exporting would corrupt the snapshot. A collision during garbage collection would crash Forest. The consequences are pretty drastic. We've only gotten away with it because collisions are so rare.

+25% peak memory usage is rough but still significantly lower than the memory required to import a snapshot (8-9 GiB).

File: src/ipld/cid_hashset.rs [hunk 1]
```diff
@@ -2,27 +2,84 @@
...
+
+// Nearly all Filecoin CIDs are V1, DagCbor encoded, and hashed with Blake2b265. If other types of
+// CID become popular, they can be added to the CidVariant structure.
+enum CidVariant {
+    V1DagCborBlake2b([u8; 32]),
+}
+
+impl TryFrom<Cid> for CidVariant {
+    type Error = ();
+    fn try_from(cid: Cid) -> Result<Self, Self::Error> {
```
Comment (LesnyRumcajs) [inline src/ipld/cid_hashset.rs:19]: Could we have a unit test for this?

File: src/ipld/cid_hashset.rs [hunk 2]
```diff
@@ -2,27 +2,84 @@
 // SPDX-License-Identifier: Apache-2.0, MIT
 
 use ahash::HashSet;
-use cid::Cid;
+use cid::{Cid, Version};
+use fvm_ipld_encoding::DAG_CBOR;
+use std::collections::BTreeSet;
 
+const BLAKE2B256: u64 = 0xb220;
```
Comment (LesnyRumcajs) [inline src/ipld/cid_hashset.rs:9]: Can we import this constant from somewhere?

File: src/ipld/cid_hashset.rs [hunk 3]
```diff
@@ -2,27 +2,84 @@
...
+
+// Nearly all Filecoin CIDs are V1, DagCbor encoded, and hashed with Blake2b265. If other types of
+// CID become popular, they can be added to the CidVariant structure.
+enum CidVariant {
+    V1DagCborBlake2b([u8; 32]),
```
Comment (LesnyRumcajs) [inline src/ipld/cid_hashset.rs:14]: `32` could be a named constant given that it appears a few times. Alternatively, the entire `[u8; 32]` could be aliased.

**Review by LesnyRumcajs — APPROVED** [review]
> > This seems like a significant drop in performance. Are you positive we are fine with this tradeoff? What would be the result of a collision, in practice?
> 
> A collision during snapshot exporting would corrupt the snapshot. A collision during garbage collection would crash Forest. The consequences are pretty drastic. We've only gotten away with it because collisions are so rare.
> 
> +25% peak memory usage is rough but still significantly lower than the memory required to import a snapshot (8-9 GiB).

The consequences are rather convincing.

File: src/ipld/cid_hashset.rs (hunk 2 above)
Comment (lemmih) [reply src/ipld/cid_hashset.rs:9]: It appears to be hidden.

File: src/ipld/cid_hashset.rs (hunk 3 above)
Comment (lemmih) [reply src/ipld/cid_hashset.rs]: Hoisted out the size of the blake2b hash.

**Pull request merged**